core/src/main/scala/org/apache/spark/MapOutputTracker.scala

```diff
@@ -827,6 +848,13 @@ private[spark] class MapOutputTrackerWorker(conf: SparkConf) extends MapOutputTr
     }
   }
 
+  override def getAllMapOutputStatuses(shuffleId: Int): Array[MapStatus] = {
+    logDebug(s"Fetching all output statuses for shuffle $shuffleId")
+    val statuses = getStatuses(shuffleId, conf)
```

Comment:
Please clean up the `mapStatuses` in case of `MetadataFetchFailedException`!

Reasoning:
The `getStatuses` method before this PR was only used in `getMapSizesByExecutorId ` where the `MetadataFetchFailedException` (the case when missing output location was detected) handled by cleaning of the `mapStatuses` cache as it is probably outdated.


I am sure that cleaning would not be missed if this cleaning would be done at the throwing of that exception.
Could you please check whether it can be moved there?

Reply:
Now I see why you cannot move the clearing there! 
Still the clearing itself is needed to be done.